### PR TITLE
fix: Project Cloning Broken for Share Links

### LINKED ISSUE #4068: Project Cloning Broken for Share Links
When I try cloning a project that is shared to me, and that I was able to clone prior to the new URL structure, I get the following error:

```json
[
    {
        "error": {
            "message": "The user must be authenticated to clone the project",
            "code": -32603,
            "data": {
                "code": "INTERNAL_SERVER_ERROR",
                "httpStatus": 500,
                "path": "project.clone"
            }
        }
    }
]
```

https://github.com/user-attachments/assets/f6285e8b-a0d6-4250-9cbd-9979d5245961

### PR DESCRIPTION
**Pull request by istarkov**
## Description

closes #4068

image

AuthorizationContext - splitted on 4 types, user, token, service, anonymous (before was a mix making it hard to understand permissions)

Clone - needs 2 contexts source and destination.

## Steps for reproduction

- **Log in as "Vasya Pupkin".** 
  `5de6:vasya@pupkin.com`
- **Create a project.**
- **Generate a ShareLink.**
- **Copy the ShareLink.**
  `https://p-636d568a-0a42-43d7-9509-8b23607cd5e8-dot-clone-project.development.webstudio.is/?authToken=da7ef507-c91b-4393-a1fe-91848ba533c9&mode=preview`
- **Log out.**

---

- **Log in as another user.**
- **Paste the ShareLink then clone the project.**
- **See title etc in clone dialog**
  image
- **Log out.**

---

- **Paste the ShareLink then clone the project.**
- **Login process**
- **See title etc in clone dialog after the login**
  image
- **Log out.**

---

- **Paste ShareLink in the, go to the menu and copy the clone link.**
  image

  `https://clone-project.development.webstudio.is/dashboard?projectToCloneAuthToken=da7ef507-c91b-4393-a1fe-91848ba533c9`

---

- **Log in as "Vasya Pupkin".**
- **Remove cloning permission for the project.**
- **Log out.**

---

Paste `https://clone-project.development.webstudio.is/dashboard?projectToCloneAuthToken=da7ef507-c91b-4393-a1fe-91848ba533c9`
See  "unexpected server error"

---

## Code Review

- [ ] hi @kof, I need you to do
  - conceptual review (architecture, feature-correctness)
  - detailed review (read every line)
  - test it on preview

## Before requesting a review

- [ ] made a self-review
- [ ] added inline comments where things may be not obvious (the "why", not "what")

## Before merging

- [ ] tested locally and on preview environment (preview dev login: 5de6)
- [ ] updated [test cases](https://github.com/webstudio-is/webstudio/blob/main/apps/builder/docs/test-cases.md) document
- [ ] added tests
- [ ] if any new env variables are added, added them to `.env` file

### CODE REVIEW

File: apps/builder/app/dashboard/dashboard.tsx [hunk 1]
```diff
@@ -49,6 +52,44 @@ export type DashboardProps = {
...
   publisherHost: string;
   imageBaseUrl: string;
+  projectToClone?: {
+    authToken: string;
+    id: string;
+    title: string;
+  };
+};
+
+const CloneProject = ({
```
[the remark this answers is not in the document]
Comment (kof) [reply apps/builder/app/dashboard/dashboard.tsx:62]: I think this should work exactly the same way as the duplicate function basically

File: apps/builder/app/builder/features/topbar/clone.tsx [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+      data-state="auto"
+      className={buttonStyle({
+        color: "positive",
+      })}
+      href={cloneProjectUrl({
+        origin: window.origin,
+        sourceAuthToken: authToken,
+      })}
+    >
+      <Text css={{ pointerEvents: "none" }} variant={"labelsSentenceCase"}>
```
[the remark this answers is not in the document]
Comment (istarkov) [reply apps/builder/app/builder/features/topbar/clone.tsx:25]: It don't, same as inside button component. For the cursor to be "buttonish"

File: apps/builder/app/dashboard/dashboard.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (istarkov) [reply apps/builder/app/dashboard/dashboard.tsx:62]: > why are we not reusing the component for Duplicate project
No consent

File: apps/builder/app/builder/features/topbar/clone.tsx [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+export const CloneButton = () => {
+  const authTokenPermission = useStore($authTokenPermissions);
+  const authToken = useStore($authToken);
+
+  if (authToken === undefined || false === authTokenPermission.canClone) {
+    return;
+  }
+
+  return (
+    <a
```
Comment (kof) [inline apps/builder/app/builder/features/topbar/clone.tsx:15]: <Link
                      className={buttonStyle({ color: "gradient" })}
                      color="contrast"
                      underline="none"
                      href="https://webstudio.is/pricing"
                      target="_blank"
                    >
                      Upgrade
                    </Link>

**Review by kof — APPROVED** [review]

**Pull request merged**